crates/matrix-sdk/src/event_cache/mod.rs

```diff
@@ -658,34 +686,29 @@ impl RoomEventCacheInner {
     /// This will only run one request; since a backpagination may need to
     /// continue, it's preferable to use [`Self::backpaginate_until`].
     ///
-    /// Returns the number of messages received in this chunk.
+    /// # Errors
+    ///
+    /// This can throw an error if we're already back-paginating.
     #[instrument(skip(self))]
-    async fn backpaginate(
-        &self,
-        batch_size: u16,
-        token: Option<PaginationToken>,
-    ) -> Result<BackPaginationOutcome> {
+    async fn paginate_backwards(&self, batch_size: u16) -> Result<BackPaginationOutcome> {
         // Make sure there's at most one back-pagination request.
-        let _guard = self.pagination_lock.lock().await;
-
-        // Get messages.
-        let messages = self
-            .room
-            .messages(assign!(MessagesOptions::backward(), {
-                from: token.as_ref().map(|token| token.0.clone()),
-                limit: batch_size.into()
-            }))
-            .await
-            .map_err(EventCacheError::SdkError)?;
+        let prev_token = self.get_or_wait_for_pagination_token().await;
+
+        self.pagination.paginator.set_idle_state(prev_token.clone(), None).await?;
+
+        // Run the actual pagination.
+        let PaginationResult { events, hit_end_of_timeline: reached_start } =
+            self.pagination.paginator.paginate_backward(batch_size.into()).await?;
 
         // Make sure the `RoomEvents` isn't updated while we are saving events from
         // backpagination.
         let mut room_events = self.events.write().await;
 
-        // Check that the `token` exists if any.
-        let gap_identifier = if let Some(token) = token.as_ref() {
+        // Check that the previous token still exists; otherwise it's a sign that the
+        // timeline has been cleared..
```

Comment:
```diff
- // timeline has been cleared..
+ // timeline has been cleared.
```

Also, it may not be appropriate to talk about “timeline” here, thoughts?

Reply:
(thanks re: typo!)

> Also, it may not be appropriate to talk about “timeline” here, thoughts?

But it's the most precise word; what would you use?